File: backend/controlstock/src/main/java/com/ues/controlstock/service/RoleService.java
package com.ues.controlstock.service;

import java.util.List;
import java.util.Optional;
import java.util.stream.Collectors;

import org.springframework.stereotype.Service;

import com.ues.controlstock.dto.RoleDTO;
import com.ues.controlstock.entity.Role;
import com.ues.controlstock.repository.RoleRepository;

// Lógica de negocio para la gestión de roles
@Service
public class RoleService {

    private final RoleRepository roleRepository;

    public RoleService(RoleRepository roleRepository) {
        this.roleRepository = roleRepository;
    }

    // Crea un nuevo rol
    public RoleDTO save(RoleDTO dto) {
        Role role = new Role();
        role.setRoleName(dto.getRoleName());
        role.setDescription(dto.getDescription());

        Role savedRole = roleRepository.save(role);
        dto.setRoleId(savedRole.getRoleId());

        return dto;
    }

    // Retorna un rol específico por su ID
    public Optional<RoleDTO> get(Long id) {
        return roleRepository.findById(id)
            .map(role -> new RoleDTO(
                role.getRoleId(),
                role.getRoleName(),
                role.getDescription()
            ));
    }

    // Retorna todos los roles registrados
    public List<RoleDTO> findAll() {
        return roleRepository.findAll()
            .stream()
            .map(this::convertToDTO)
            .collect(Collectors.toList());
    }

    // Actualiza un rol existente
    public RoleDTO update(Long id, RoleDTO updatedData) {
        Role role = roleRepository.findById(id).orElseThrow();
        role.setRoleName(updatedData.getRoleName());
        role.setDescription(updatedData.getDescription());
        return convertToDTO(roleRepository.save(role));
    }

    // Elimina un rol por su ID — retorna true si fue eliminado
    public boolean delete(Long id) {
        if (roleRepository.existsById(id)) {
            roleRepository.deleteById(id);
            return true;
        }
        return false;
    }

    // Convierte la entidad Role a DTO
    private RoleDTO convertToDTO(Role role) {
        return new RoleDTO(
            role.getRoleId(),
            role.getRoleName(),
            role.getDescription()
        );
    }
}

File: frontend/controlstock/src/core/api/services/role-service.ts
import type { Role } from "../../entity";
import { httpService } from "../http-service";

const ENDPOINT = "/roles";

/**
 * Servicio para operaciones CRUD de Roles contra la API REST.
 * Reemplaza la simulación en localStorage de utils/roleCrud.ts
 */
export const roleService = {
  /**
   * Obtiene todos los roles.
   * GET /api/roles
   */
  getAll() {
    return httpService.get<Role[]>(ENDPOINT);
  },

  /**
   * Obtiene un rol por su ID.
   * GET /api/roles/:id
   */
  getById(id: number) {
    return httpService.get<Role>(`${ENDPOINT}/${id}`);
  },

  /**
   * Crea un nuevo rol.
   * POST /api/roles
   */
  create(role: Omit<Role, "id">) {
    return httpService.post<Role>(ENDPOINT, role);
  },

  /**
   * Actualiza un rol existente.
   * PUT /api/roles/:id
   */
  update(id: number, role: Omit<Role, "id">) {
    return httpService.put<Role>(`${ENDPOINT}/${id}`, role);
  },

  /**
   * Elimina un rol.
   * DELETE /api/roles/:id
   */
  delete(id: number) {
    return httpService.delete<void>(`${ENDPOINT}/${id}`);
  },
};